src/mesh/NextHopRouter.cpp

```diff
@@ -0,0 +1,106 @@
+#include "NextHopRouter.h"
+
+NextHopRouter::NextHopRouter() {}
+
+/**
+ * Send a packet
+ */
+ErrorCode NextHopRouter::send(meshtastic_MeshPacket *p)
+{
+    // Add any messages _we_ send to the seen message list (so we will ignore all retransmissions we see)
+    wasSeenRecently(p); // FIXME, move this to a sniffSent method
+
+    p->next_hop = getNextHop(p->to, p->current_relayer); // set the next hop
+    LOG_DEBUG("Setting next hop for packet with dest %x to %x\n", p->to, p->next_hop);
+
+    return Router::send(p);
+}
+
+bool NextHopRouter::shouldFilterReceived(const meshtastic_MeshPacket *p)
+{
+    if (wasSeenRecently(p)) { // Note: this will also add a recent packet record
+        if (p->next_hop == getNodeNum()) {
+            LOG_DEBUG("Ignoring incoming msg, because we've already seen it.\n");
+        } else {
+            LOG_DEBUG("Ignoring incoming msg, because we've already seen it and cancel any outgoing packets.\n");
+            Router::cancelSending(p->from, p->id);
+        }
+        return true;
+    }
+
+    return Router::shouldFilterReceived(p);
+}
+
+void NextHopRouter::sniffReceived(const meshtastic_MeshPacket *p, const meshtastic_Routing *c)
+{
+    bool isAck =
+        ((c && c->error_reason == meshtastic_Routing_Error_NONE)); // consider only ROUTING_APP message without error as ACK
+    if (isAck) {
+        // Update next-hop of this successful transmission to current relayer, but ONLY if "from" is not 0 or ourselves (means
+        // implicit ACK or someone is relaying our ACK)
+        if (p->from != 0 && p->from != getNodeNum()) {
+            if (p->current_relayer) {
+                meshtastic_NodeInfoLite *sentTo = nodeDB.getMeshNode(p->from);
+                if (sentTo) {
+                    LOG_DEBUG("Update next hop of %x to %x based on received ACK.\n", p->from, p->current_relayer);
+                    sentTo->next_hop = p->current_relayer;
+                }
+            }
+        }
+    }
+
+    if (config.device.role != meshtastic_Config_DeviceConfig_Role_CLIENT_MUTE) {
+        if ((p->to != getNodeNum()) && (getFrom(p) != getNodeNum())) {
+            if (p->next_hop == getNodeNum()) {
```

Comment:
If a function is added (part of NeighborInfoModule?) that determines if we only have one neighbor that could help us avoid "tip of tree" rebroadcasts. For example:

if (p->next_hop == getNodeNum() && !hasOnlyOneNeighbor()) {